File: 10-leetcodes-entrevistas/python/two-pointer.py
# Exercise: Valid Palindrome

class Solution:
  def isPalindrome(self, s: str) -> bool:
    l = 0
    r = len(s) -1

    while l < r:
      if not s[l].isalnum():
        l+=1
      elif not s[r].isalnum():
        r-=1
      elif s[l].lower() == s[r].lower():
        l+=1
        r-=1
      else:
        return False
    return True

s = Solution()
print(s.isPalindrome("A man, a plan, a canal: Panama"))
print(s.isPalindrome("race a car"))

File: leetcodes-para-entrevistas/php/two-pointer.php
<?php

declare(strict_types=1);

// Exercise: Valid Palindrome

class Solution
{
  public function isPalindrome(string $s): bool
  {
    $left = 0;
    $right = strlen($s) - 1;

    while ($left < $right) {
      if (!ctype_alpha($s[$left]) && $left < $right) {
        $left++;
      } else if (!ctype_alpha($s[$right]) && $left < $right) {
        $right--;
      } else if (strtolower($s[$left]) === strtolower($s[$right])) {
        $left++;
        $right--;
      } else {
        return false;
      }
    }

    return true;
  }
}

$solution = new Solution();
echo $solution->isPalindrome('A man, a plan, a canal: Panama') . PHP_EOL;
